clippy2021.md

```diff
@@ -0,0 +1,178 @@
+# Clippy Roadmap 2021
+
+As the Rust language gets bigger and bigger, also Clippy gets more and more
+attention. This is good for the project, as it gets more and more users and
+contributors. With that we are seeing more and more issues about the reliability
+and usability of Clippy. Some examples are:
+
+- `cargo clippy` after `cargo check` is the most
+  prominent and has many duplicate issues referring to it:
+  [#4612](https://github.com/rust-lang/rust-clippy/issues/4612)
+- Better testing of clippy lints before enabling them by default:
+  [#6429](https://github.com/rust-lang/rust-clippy/issues/6429)
+- Add an easy way to run clippy with high FPR lints disabled:
+  [#5537](https://github.com/rust-lang/rust-clippy/issues/5537)
+- Missing warn-by-default correctness group:
+  [#6366](https://github.com/rust-lang/rust-clippy/issues/6366)
+
+And many more, that are either duplicates or address other things about
+reliability and usability.
+
+Besides that, there are internal things that aren't optimal currently. Those
+span from Clippy development to management of Clippy. This document will list
+things, we should address in 2021 and should start a conversation how we should
+prioritize things. It is split up into external and internal issues, which are
+further split up.
+
+## External
+
+This section lists things that are user facing. We should make Clippy as
+pleasant to use as possible. This section is split up in [usability](#Usability)
+and [reliability](#Reliability).
+
+### Usability
+
+This addresses issues we're seeing around usability of Clippy. Clippy should be
+easy to use and easy to configure.
+
+#### `cargo clippy` after `cargo check`
+
+Currently when `clippy` is run after `cargo check`, it does not produce any
+output. This is especially problematic since `rust-analyzer` is on the rise and
+it uses `cargo check` for checking code. A fix is already implemented, but it
+still has to be pushed over the finish line.
+
+[#4612](https://github.com/rust-lang/rust-clippy/issues/4612)
+
+#### Lint groups
+
+The are more and more issues about managing lints in Clippy. Lints are hard to
+implement with a guarantee of no/few false positives (FPs). This won't change.
+But we should consider implementing more lint groups or improve the process of
+classifying lints. This will also involve internal changes, and will be
+addressed later in this document again.
+
+[#5537](https://github.com/rust-lang/rust-clippy/issues/5537)
+[#6366](https://github.com/rust-lang/rust-clippy/issues/6366)
+
+#### `lints.toml` configuration
+
+This is something that comes up every now or then. A configuration file, where
+lint levels can be defined and reused across projects. Discussions about this
+often lead to nothing specific/actionable or to "we need an RFC for this". And
+this is exactly what we need to do. Get together with the cargo team and write
+an RFC and implement such a configuration file somehow and somewhere.
```

Comment:
I can try to write up a summary in the tracking issue if that's alright

Reply:
I don't think we get far with only discussing this in issues. There were other concerns, like where to implement it: Should it be an extra file? Should it live in `Cargo.toml`? Or a hybrid version, so that you can link a file in the `Cargo.toml` but can also put it directly in there. Also the reproducibility of compiling probably needs more discussion.

I don't think this can be done without an RFC (in a joint effort of cargo+clippy).

Reply:
At least the matter of files seems to have been decided by the cargo team ([see this IRLO comment](https://internals.rust-lang.org/t/proposal-cargo-lint-configuration/9135/6))

But yes, an RFC is absolutely needed =)

Reply:
> At least the matter of files seems to have been decided by the cargo team

I don't know, if that's really decided. I think the cargo team would really like for it to go into `Cargo.toml` but most people find a extra file for this more intuitive. That's at least the impression I got from closing the issues about this feature in the `clippy.toml` file and other people I talked to about this.